File: muxer/src/events.py
"""Redis Pub/Sub event publishing for Muxing Worker."""

import json
from typing import Any

import redis
import structlog

from .config import config
from .types import EventMessage

logger = structlog.get_logger(__name__)


class EventPublisher:
    """Publishes events to Redis Pub/Sub channels."""

    def __init__(self, redis_client: redis.Redis):
        self.redis = redis_client

    def publish(self, channel: str, message: EventMessage) -> None:
        """Publish an event message to a channel."""
        try:
            payload = json.dumps(
                {
                    "jobId": message.job_id,
                    "type": message.type,
                    "timestamp": message.timestamp,
                    "payload": message.payload,
                }
            )
            self.redis.publish(channel, payload)
        except Exception as e:
            logger.error("Failed to publish event", channel=channel, error=str(e))

    def publish_progress(self, job_id: str, stage: str, percent: float) -> None:
        """Publish a progress event."""
        payload: dict[str, Any] = {
            "stage": stage,
            "percent": percent,
        }
        message = EventMessage.create(job_id, "progress", payload)
        self.publish("events:progress", message)

    def publish_state_change(self, job_id: str, from_status: str, to_status: str) -> None:
        """Publish a state change event."""
        payload = {"from": from_status, "to": to_status}
        message = EventMessage.create(job_id, "state_change", payload)
        self.publish("events:state", message)

    def publish_log(self, job_id: str, level: str, log_message: str) -> None:
        """Publish a log event."""
        payload = {"level": level, "message": log_message}
        message = EventMessage.create(job_id, "log", payload)
        self.publish("events:log", message)

    def publish_error(
        self,
        job_id: str,
        code: str,
        error_message: str,
        retryable: bool,
        stack: str | None = None,
    ) -> None:
        """Publish an error event."""
        payload: dict[str, Any] = {
            "code": code,
            "message": error_message,
            "retryable": retryable,
        }
        if stack:
            payload["stack"] = stack

        message = EventMessage.create(job_id, "error", payload)
        self.publish("events:error", message)


def create_event_publisher() -> EventPublisher:
    """Create an event publisher with Redis connection."""
    redis_client = redis.Redis.from_url(config.redis_url)
    return EventPublisher(redis_client)

File: dubber/src/events.ts
import Redis from 'ioredis';
import { config } from './config.js';
import type { EventMessage } from './types.js';

let redis: Redis | null = null;
let redisReady = false;

export function getRedis(): Redis {
  if (!redis) {
    redis = new Redis(config.redisUrl, {
      maxRetriesPerRequest: null,
      retryStrategy(times: number) {
        // Exponential backoff with max 30 seconds
        const delay = Math.min(times * 1000, 30000);
        console.log(`Redis reconnecting in ${delay}ms (attempt ${times})`);
        return delay;
      },
    });

    redis.on('connect', () => {
      console.log('Redis events client connected');
      redisReady = true;
    });

    redis.on('error', (err) => {
      console.error('Redis events client error:', err.message);
      redisReady = false;
    });

    redis.on('close', () => {
      console.log('Redis events client disconnected');
      redisReady = false;
    });

    redis.on('reconnecting', () => {
      console.log('Redis events client reconnecting...');
    });
  }
  return redis;
}

export function isRedisReady(): boolean {
  return redisReady;
}

export async function closeRedis(): Promise<void> {
  if (redis) {
    await redis.quit();
    redis = null;
  }
}

export async function publishEvent(channel: string, message: EventMessage): Promise<void> {
  try {
    await getRedis().publish(
      channel,
      JSON.stringify({
        jobId: message.jobId,
        type: message.type,
        timestamp: message.timestamp,
        payload: message.payload,
      })
    );
  } catch (err) {
    console.error(`Failed to publish event to ${channel}:`, err);
  }
}

export async function publishProgress(
  jobId: string,
  stage: string,
  percent: number
): Promise<void> {
  await publishEvent('events:progress', {
    jobId,
    type: 'progress',
    timestamp: new Date().toISOString(),
    payload: { stage, percent },
  });
}

export async function publishStateChange(
  jobId: string,
  fromStatus: string,
  toStatus: string
): Promise<void> {
  await publishEvent('events:state', {
    jobId,
    type: 'state_change',
    timestamp: new Date().toISOString(),
    payload: { from: fromStatus, to: toStatus },
  });
}

export async function publishLog(
  jobId: string,
  level: string,
  message: string
): Promise<void> {
  await publishEvent('events:log', {
    jobId,
    type: 'log',
    timestamp: new Date().toISOString(),
    payload: { level, message },
  });
}

export async function publishError(
  jobId: string,
  code: string,
  message: string,
  retryable: boolean,
  stack?: string
): Promise<void> {
  await publishEvent('events:error', {
    jobId,
    type: 'error',
    timestamp: new Date().toISOString(),
    payload: { code, message, retryable, stack },
  });
}
